File: Baekjoon/Class4/18.13549 숨바꼭질 3.cpp
#include <iostream>
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, k;
    cin >> n >> k;
    int depth_limit = max(n, k) * 2;

    deque<int> dq; // 0-1 너비 우선 탐색
    vector<int> depths(depth_limit + 1, -1);
    dq.push_back(n); // 시작
    depths[n] = 0;

    while (!dq.empty()) {
        int start = dq.front();
        if (start == k) break;
        dq.pop_front();

        int end = start * 2;
        if (end <= depth_limit && depths[end] == -1) { // 순간 이동
            dq.push_front(end);
            depths[end] = depths[start];
        }

        for (int end : {start - 1, start + 1}) // 걷기
            if (end >= 0 && end <= depth_limit && depths[end] == -1) {
                dq.push_back(end);
                depths[end] = depths[start] + 1;
            }
    }

    cout << depths[k];
    return 0;
}

File: Baekjoon/Class4/18.13549 숨바꼭질 3.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Deque;
import java.util.StringTokenizer;

class Main {
    public static void main(String[] args) throws IOException {
        int n, k;
        int depthLimit;
        Deque<Integer> dq = new ArrayDeque<>();
        int[] depths;
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        n = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        depthLimit = Math.max(n, k) * 2;

        depths = new int[depthLimit + 1]; // 0-1 너비 우선 탐색
        Arrays.fill(depths, -1);
        dq.addLast(n); // 시작
        depths[n] = 0;

        while(!dq.isEmpty()) {
            int start = dq.removeFirst();
            if(start == k) break;

            int teleport = start * 2;
            if(teleport <= depthLimit && depths[teleport] == -1) { // 순간 이동
                dq.addLast(teleport);
                depths[teleport] = depths[start];
            }

            for(int walk: new int[]{ start - 1, start + 1 }) // 걷기
                if(walk >= 0 && walk <= depthLimit && depths[walk] == -1) {
                    dq.addLast(walk);
                    depths[walk] = depths[start] + 1;
                }
        }

        System.out.print(depths[k]);
        br.close();
    }
}